### PR TITLE
tested all functions in heiferI.py

### PR DESCRIPTION
**Pull request by 0luchi**
Added test cases in test_animal_life_cycle.py for all the functions in heiferI.py
- Used mock.MagicMock() arguments
- Test cases are in test_heiferI() and test_update_heiferI()
- Corrected an error in line 58 of heiferI.py

### CODE REVIEW

**Comment by ew3361zh** [comment]
I haven't done a full review yet but wanted to ask a couple questions first:

1. Is this addressing a particular issue to increase coverage for heiferI.py? If so, could you link the issue to this PR to ensure we know the full extent of the issue being addressed?
2. This looks like it increases test coverage for heiferI.py greatly from 27% (currently on main) to 86% (on this branch). Did you look at covering the other lines (35, 43, 45, 47-49, 84-85) to get to 100%?

**Comment by 0luchi** [comment]
@ew3361zh 
1. I don't think it is addressing an issue. I was just working based on this spreadsheet https://docs.google.com/spreadsheets/d/1wjdPo_20AtT84-809bf_-FXA8h6e00zMk8prx7_d3vc/edit?pli=1#gid=0
2. I will work on covering those lines. This is my first PR so I wanted to make sure I was doing the testing correctly first.

**Comment by 0luchi** [comment]
@ew3361zh 
I just improved coverage to 100%

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (JoeWaddell) [inline tests/animal_module_tests/test_animal_life_cycle.py]: This needs to be the `AnimalGroupingScenario` class, not the `AnimalType` itself! Likewise this change will require the rolling back of the change in `heifer1.py`

File: RUFAS/routines/animal/life_cycle/heiferI.py
Comment (JoeWaddell) [inline RUFAS/routines/animal/life_cycle/heiferI.py]: This should be rolled back, since the methods in `animal_requirements.py` need the AnimalGroupingScenario type, not AnimalType directly. It'ssomething we can consider refactoring, if you see a problem its causing here.

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (JoeWaddell) [inline tests/animal_module_tests/test_animal_life_cycle.py]: Each of these separate methods should be separated into different tests. It will increase the length of this file a little, but will be easier to maintain tests as methods are added/removed, since it'll ensure that individual tests aren't reliant on things that happened in other tests.

**Review by JoeWaddell — CHANGES_REQUESTED** [review]
This is looking good! I've left some comments, mostly about breaking it up into separate unit tests, for maintainability. Should be straightforward, though it will lead to a longer file.

Also a couple questions about a change to the logic in heifer1.py, which I think will need to be reverted (and may require rewriting a couple sections of the tests).

File: RUFAS/routines/animal/life_cycle/heiferI.py
Comment (0luchi) [reply RUFAS/routines/animal/life_cycle/heiferI.py]: Sorry about that. I was just a little confused with what the type of animal_grouping_scenario is because the calc_rqmts() function had the type annotation of AnimalType, but this clarifies things now thank you.

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (JoeWaddell) [inline tests/animal_module_tests/test_animal_life_cycle.py]: For each unit test here you'll want a docstring summary that matches the format of the others: in this case 

`Unit test for method set_nutrient_rqmts in file heiferI.py.`

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (JoeWaddell) [inline tests/animal_module_tests/test_animal_life_cycle.py]: Here I think you could add an `assert_called_with` to check that it used the expected values in the call.

Additionally, I think you could add another assertion here, namely that the mocked heiferI object correctly applies the requirement values from the animal_requirements object.

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (JoeWaddell) [inline tests/animal_module_tests/test_animal_life_cycle.py:1555]: I think a check that `manure_calculations` was also called would be good here.

**Review by JoeWaddell — CHANGES_REQUESTED** [review]
Looking good! Just had a couple comments on docstring formatting and a couple extra assertions needed.

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (0luchi) [reply tests/animal_module_tests/test_animal_life_cycle.py:1555]: I tried doing this, but I've been having trouble because manure_calculations does not belong to a class. I'm not sure of how to mock it

Comment (JoeWaddell) [reply tests/animal_module_tests/test_animal_life_cycle.py:1555]: The confusing thing is that since Heifer I is mocked, we have to mock manure_calculations where it's imported, instead of from where it 'should' exist.
So it can be mocked as follows:
`thingy = mocker.patch('RUFAS.routines.animal.life_cycle.heiferI.manure_calculations', return_value=(None, None))`
That mocked method can then be checked via things like: `thingy.assert_called_once()`, and even better, `assert_called_with`, etc.
Of course we'd want better names than 'thingy'.

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (Isgotkowitz) [inline tests/animal_module_tests/test_animal_life_cycle.py]:
```suggestion
```

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (Isgotkowitz) [inline tests/animal_module_tests/test_animal_life_cycle.py]: I think it would be good to define the value `(15.625 * 0.64)/10.0` (it's just 1.0) in a variable and label it. Otherwise it is very confusing where these numbers are coming from. The way that `net_energy_diet_concentration` is calculated is also very weird - there are multiple paths of execution but if the defaults for `metabolizable_energy` and `previous_DMI` are used, then both paths lead to the same result. I think it would be good to pass in non-default and non-zero values for `metabolizable_energy` and `previous_DMI`, and check that the `net_energy_diet_concentration` is calculated to be what you expect it to be.

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (Isgotkowitz) [reply tests/animal_module_tests/test_animal_life_cycle.py:1555]: Also, `self.p_excrt` and `self.manure_excretion` should be checked to make sure they were set correctly.

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (Isgotkowitz) [inline tests/animal_module_tests/test_animal_life_cycle.py]: This is entirely duplicating the logic of `phosphorus_rqmts`, which means that if there are any bugs in the method they will be propagated into the test. The values that you expect to be calculated for `p_maint_feces`, `p_growth`, and `p_req` should be hardcoded in the test (in an ideal world - in practice it can be difficult and sometimes impractical to do this for large unwieldy math equations).

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (Isgotkowitz) [inline tests/animal_module_tests/test_animal_life_cycle.py]:
```suggestion
    mocker.patch('RUFAS.routines.animal.life_cycle.life_cycle.AnimalBase.config', animal_base_config)
    
    actual_weight = HeiferI.get_non_preg_bw_change(heiferI)
    
    expected_weight = 0.55 * 0.96 * heiferI.mature_body_weight - 0.96 * heiferI.body_weight  # divisor == 1
    assert actual_weight == expected_weight
```
Grouping by setup, act, and assert helps to keep things more readable. Also, using the terms actual and expected for what the method actually does vs what you expect it to do also makes the test a lot more readable.

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (Isgotkowitz) [inline tests/animal_module_tests/test_animal_life_cycle.py]: This test is only testing a single case: when `days_born` is exactly equal to `AnimalBase.config["target_heifer_preg_day"]`. It should be checking the other two cases, when days born is > and < the target heifer preg day. All that said, I think you should hold off on expanding this test until talking w/ @JoeWaddell about this method, because he is going to evaluate this method's logic to see if it needs to be changed.

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (Isgotkowitz) [inline tests/animal_module_tests/test_animal_life_cycle.py:1640]: It would be good to set the return value for `get_non_preg_bw_change()` and check that daily growth is being correctly set.

File: tests/animal_module_tests/test_animal_life_cycle.py
Comment (Isgotkowitz) [inline tests/animal_module_tests/test_animal_life_cycle.py]: The body weight change should be checked in this test as well.

**Review by Isgotkowitz — APPROVED** [review]
Overall this PR looks good. Several comments about improving tests, not all suggestions for improvements may be feasible though. After looking through those (and getting a second approval) I think this PR should be good to go.

Keep up the good work Oluchi!

**Review by JoeWaddell — APPROVED** [review]
Nice job updating these after all the comments! LGTM

**Pull request merged**
